Migrate tests from "api4/team_test.go" to use testify

gupsho opened the pull request:
#### Summary
Migrate tests from "api4/team_test.go" to use testify

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/12882

hanzei commented:
The build is still failing: https://mattermost.atlassian.net/secure/RapidBoard.jspa?rapidView=35&projectKey=MM&view=planning

hahmadia commented on api4/team_test.go:
```diff
@@ -2364,42 +2124,29 @@ func TestInviteUsersToTeam(t *testing.T) {
-			t.Fatal("Should update the team")
-		}
+		_, err = th.App.UpdateTeam(th.BasicTeam)
+		require.Nil(t, err, "Should update the team")
```
```suggestion
		require.Nilf(t, err, "%v, Should update the team", err)
```

hahmadia commented on api4/team_test.go:
```diff
@@ -2364,42 +2124,29 @@ func TestInviteUsersToTeam(t *testing.T) {
-			t.Fatal("Per team restriction should take precedence over the global restriction")
-		}
+		err = th.App.InviteNewUsersToTeam([]string{"test@global.com"}, th.BasicTeam.Id, th.BasicUser.Id)
+		require.NotNil(t, err, "Per team restriction should take precedence over the global restriction")
```
```suggestion
		require.NotNilf(t, err, "%v, Per team restriction should take precedence over the global restriction", err)
```

hahmadia commented on api4/team_test.go:
```diff
@@ -2364,42 +2124,29 @@ func TestInviteUsersToTeam(t *testing.T) {
-		}
+		err = th.App.InviteNewUsersToTeam([]string{"test@global.com"}, th.BasicTeam.Id, th.BasicUser.Id)
+		require.NotNil(t, err, "Per team restriction should take precedence over the global restriction")
+		require.Equal(t, err.Where, "InviteNewUsersToTeam", "Per team restriction should take precedence over the global restriction")
```
```suggestion
		require.Equalf(t, err.Where, "InviteNewUsersToTeam", "%v, Per team restriction should take precedence over the global restriction", err)
```

hahmadia commented on api4/team_test.go:
```diff
@@ -2364,42 +2124,29 @@ func TestInviteUsersToTeam(t *testing.T) {
-			t.Fatal("Should not invite user")
-		}
+		err = th.App.InviteNewUsersToTeam([]string{"test@common.com"}, th.BasicTeam.Id, th.BasicUser.Id)
+		require.Nil(t, err, "Failed to invite user which was common between team and global domain restriction")
```
```suggestion
		require.Nilf(t, err, "%v, Failed to invite user which was common between team and global domain restriction", err)
```

hahmadia commented on api4/team_test.go:
```diff
@@ -2364,42 +2124,29 @@ func TestInviteUsersToTeam(t *testing.T) {
+		require.Nil(t, err, "Failed to invite user which was common between team and global domain restriction")
 
+		err = th.App.InviteNewUsersToTeam([]string{"test@invalid.com"}, th.BasicTeam.Id, th.BasicUser.Id)
+		require.NotNil(t, err, "Should not invite user")
```
```suggestion
		require.NotNilf(t, err, "%v, Should not invite user", err)
```

hahmadia commented on api4/team_test.go:
```diff
@@ -2553,22 +2284,18 @@ func TestSetTeamIcon(t *testing.T) {
-	if err != nil {
-		t.Fatal(err)
-	}
+	require.Nil(t, err)
```
```suggestion
	require.Nil(t, err, err)
```

hahmadia commented on api4/team_test.go:
```diff
@@ -2604,9 +2331,8 @@ func TestSetTeamIcon(t *testing.T) {
-		t.Fatal(err)
-	}
+	err = th.cleanupTestFile(info)
+	require.Nil(t, err)
```
```suggestion
	require.Nil(t, err, err)
```

gupsho commented on api4/apitestlib.go:
```diff
@@ -591,27 +591,21 @@ func CheckEtag(t *testing.T, data interface{}, resp *model.Response) {
-	if resp.Error != nil {
-		t.Fatalf("Expected no error, got %q", resp.Error.Error())
-	}
+	require.Nilf(t, resp.Error, "Expected no error")
```
Hi @hahmadia I had it the may you are suggesting but was getting an exception when running tests. I think, when error was nil, require was still evaluating the params and throwing a NPE.

gupsho commented on api4/team_test.go:
```diff
@@ -57,15 +51,10 @@ func TestCreateTeam(t *testing.T) {
+	r, err := Client.DoApiPost("/teams", "garbage")
+	require.NotNil(t, err, "should have errored")
+
+	require.Equalf(t, r.StatusCode, http.StatusBadRequest, "wrong status code actual: %s, expected: %s", strconv.Itoa(r.StatusCode), strconv.Itoa(http.StatusBadRequest))
```
Done.

gupsho commented:
> Replied to your concerns regarding the Null Pointer Error. Great work! Thanks for addressing the problems so quickly.

Updated.

jespino commented on api4/team_test.go:
```diff
@@ -114,9 +101,7 @@ func TestCreateTeamSanitization(t *testing.T) {
-		if rteam.Email == "" {
-			t.Fatal("should not have sanitized email")
-		}
+		require.NotEqual(t, rteam.Email, "", "should not have sanitized email")
```
Probably would be more semantic here something like `require.NotEmpty`

jespino commented on api4/team_test.go:
```diff
@@ -130,9 +115,7 @@ func TestCreateTeamSanitization(t *testing.T) {
-		if rteam.Email == "" {
-			t.Fatal("should not have sanitized email")
-		}
+		require.NotEqual(t, rteam.Email, "", "should not have sanitized email")
```
Probably would be more semantic here something like `require.NotEmpty`

jespino commented on api4/team_test.go:
```diff
@@ -204,25 +185,20 @@ func TestGetTeamSanitization(t *testing.T) {
-			t.Fatal("should've sanitized email")
-		}
+
+		require.Equal(t, rteam.Email, "", "should have sanitized email")
```
Probably would be more semantic here something like `require.Empty`

jespino commented on api4/team_test.go:
```diff
@@ -204,25 +185,20 @@ func TestGetTeamSanitization(t *testing.T) {
-		if rteam.Email == "" {
-			t.Fatal("should not have sanitized email")
-		}
+		require.NotEqual(t, rteam.Email, "", "should not have sanitized email")
```
Probably would be more semantic here something like `require.NotEmpty`

jespino commented on api4/team_test.go:
```diff
@@ -204,25 +185,20 @@ func TestGetTeamSanitization(t *testing.T) {
-		if rteam.Email == "" {
-			t.Fatal("should not have sanitized email")
-		}
+		require.NotEqual(t, rteam.Email, "", "should not have sanitized email")
```
Probably would be more semantic here something like `require.NotEmpty`

jespino commented on api4/team_test.go:
```diff
@@ -384,17 +337,13 @@ func TestUpdateTeamSanitization(t *testing.T) {
-		if rteam.Email == "" {
-			t.Fatal("should not have sanitized email for admin")
-		}
+		require.NotEqual(t, rteam.Email, "", "should not have sanitized email for admin")
```
Probably would be more semantic here something like `require.NotEmpty`

jespino commented on api4/team_test.go:
```diff
@@ -384,17 +337,13 @@ func TestUpdateTeamSanitization(t *testing.T) {
-		if rteam.Email == "" {
-			t.Fatal("should not have sanitized email for admin")
-		}
+		require.NotEqual(t, rteam.Email, "", "should not have sanitized email for admin")
```
Probably would be more semantic here something like `require.NotEmpty`

jespino commented on api4/team_test.go:
```diff
@@ -489,17 +421,13 @@ func TestPatchTeamSanitization(t *testing.T) {
-		if rteam.Email == "" {
-			t.Fatal("should not have sanitized email for admin")
-		}
+		require.NotEqual(t, rteam.Email, "", "should not have sanitized email for admin")
```
Probably would be more semantic here something like `require.NotEmpty`

jespino commented on api4/team_test.go:
```diff
@@ -489,17 +421,13 @@ func TestPatchTeamSanitization(t *testing.T) {
-		if rteam.Email == "" {
-			t.Fatal("should not have sanitized email for admin")
-		}
+		require.NotEqual(t, rteam.Email, "", "should not have sanitized email for admin")
```
Probably would be more semantic here something like `require.NotEmpty`

jespino commented on api4/team_test.go:
```diff
@@ -783,20 +701,15 @@ func TestGetAllTeamsSanitization(t *testing.T) {
-				if rteam.Email == "" {
-					t.Fatal("should not have sanitized email for team admin")
-				}
+				require.NotEqual(t, rteam.Email, "", "should not have sanitized email for team admin")
```
Probably would be more semantic here something like `require.NotEmpty`

jespino commented on api4/team_test.go:
```diff
@@ -783,20 +701,15 @@ func TestGetAllTeamsSanitization(t *testing.T) {
-				if rteam.Email != "" {
-					t.Fatal("should've sanitized email for non-admin")
-				}
+				require.Equal(t, rteam.Email, "", "should've sanitized email for non-admin")
```
Probably would be more semantic here something like `require.Empty`

jespino commented on api4/team_test.go:
```diff
@@ -882,25 +791,19 @@ func TestGetTeamByNameSanitization(t *testing.T) {
-		if rteam.Email != "" {
-			t.Fatal("should've sanitized email")
-		}
+		require.Equal(t, rteam.Email, "", "should've sanitized email")
```
Probably would be more semantic here something like `require.Empty`

jespino commented on api4/team_test.go:
```diff
@@ -882,25 +791,19 @@ func TestGetTeamByNameSanitization(t *testing.T) {
-		if rteam.Email == "" {
-			t.Fatal("should not have sanitized email")
-		}
+		require.NotEqual(t, rteam.Email, "", "should not have sanitized email")
```
Probably would be more semantic here something like `require.NotEmpty`

jespino commented on api4/team_test.go:
```diff
@@ -911,71 +814,51 @@ func TestSearchAllTeams(t *testing.T) {
-	if len(rteams) != 1 {
-		t.Fatal("should have returned 1 team")
-	}
+	require.Equal(t, len(rteams), 1, "should have returned 1 team")
```
Would be more semantic `require.Len` instead of `require.Equal` + `len(rteams)`

jespino commented on api4/team_test.go:
```diff
@@ -911,71 +814,51 @@ func TestSearchAllTeams(t *testing.T) {
-	if len(rteams) != 1 {
-		t.Fatal("should have returned 1 team")
-	}
+	require.Equal(t, len(rteams), 1, "should have returned 1 team")
```
`require.Len`

jespino commented on api4/team_test.go:
```diff
@@ -911,71 +814,51 @@ func TestSearchAllTeams(t *testing.T) {
-	if len(rteams) != 0 {
-		t.Fatal("should have not returned team")
-	}
+	require.Equal(t, len(rteams), 0, "should have not returned team")
```
`require.Len`

jespino commented on api4/team_test.go:
```diff
@@ -911,71 +814,51 @@ func TestSearchAllTeams(t *testing.T) {
-	if len(rteams) != 0 {
-		t.Fatal("should have not returned team")
-	}
+	require.Equal(t, len(rteams), 0, "should have not returned team")
```
`require.Len`

jespino commented on api4/team_test.go:
```diff
@@ -911,71 +814,51 @@ func TestSearchAllTeams(t *testing.T) {
-	if len(rteams) != 1 {
-		t.Fatal("should have returned 1 team")
-	}
+	require.Equal(t, len(rteams), 1, "should have returned 1 team")
```
`require.Len`

jespino commented on api4/team_test.go:
```diff
@@ -911,71 +814,51 @@ func TestSearchAllTeams(t *testing.T) {
-	if len(rteams) != 0 {
-		t.Fatal("should have not returned team")
-	}
+	require.Equal(t, len(rteams), 0, "should have not returned team")
```
`require.Len`

jespino commented on api4/team_test.go:
```diff
@@ -1014,11 +897,8 @@ func TestSearchAllTeamsSanitization(t *testing.T) {
-			} else if rteam.AllowedDomains != "" {
-				t.Fatal("should've sanitized allowed domains")
-			}
+			require.Equal(t, rteam.Email, "", "should've sanitized email")
```
Probably would be more semantic here something like `require.Empty`

gupsho commented:
> I added a bunch of comments, but in general I would like to see usage of `require.Empty` or `require.NotEmpty` instead of `require.Equal(t, variable, "")` and `require.NotEqual(t, variable, "")`.
> 
> Besides, I would like to see `require.Len` instead of `require.Equal(t, len(variable), number)`.
> 
> I didn't add all the comments because there are a lot of this same cases.
> 
> Otherwise looks good to me. Thanks @gupsho.

Fixed.

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

hanzei commented:
@gupsho

hanzei commented:
@gupsho Would you please merge `master` into your branch?

jespino commented on api4/team_test.go:
```diff
@@ -1691,8 +1509,9 @@ func TestAddTeamMembers(t *testing.T) {
-	if err := th.App.RemoveUserFromTeam(th.BasicTeam.Id, th.BasicUser2.Id, ""); err != nil {
-		t.Fatalf(err.Error())
+	err := th.App.RemoveUserFromTeam(th.BasicTeam.Id, th.BasicUser2.Id, "")
+	if err != nil {
```
You can replace it with an `require.Nilf(t, err, "success expected, error %v found", err)`

jespino commented on api4/team_test.go:
```diff
@@ -2345,16 +2116,10 @@ func TestInviteUsersToTeam(t *testing.T) {
-				}
+			require.True(t, strings.ContainsAny(resultsMailbox[len(resultsMailbox)-1].To[0], email), "Wrong To recipient")
+			resultsEmail, err := mailservice.GetMessageFromMailbox(email, resultsMailbox[len(resultsMailbox)-1].ID)
+			if err == nil {
```
Probably here is just a `require.Nil(t, err)`

gupsho replied:
The code here checks if err is nil then resultsEmail.Subject should be equal to expectedSubject else there is no failure and we move forward. Using require.Nil will fail is err is not nil.

jespino commented on api4/team_test.go:
```diff
@@ -2479,16 +2226,10 @@ func TestInviteGuestsToTeam(t *testing.T) {
-				}
+			require.True(t, strings.ContainsAny(resultsMailbox[len(resultsMailbox)-1].To[0], email), "Wrong To recipient")
+			resultsEmail, err := mailservice.GetMessageFromMailbox(email, resultsMailbox[len(resultsMailbox)-1].ID)
+			if err == nil {
```
Probably here you can just use a `require.Nil(t, err)`

gupsho replied:
The code here checks if err is nil then resultsEmail.Subject should be equal to expectedSubject else there is no failure and we move forward. Using require.Nil will fail is err is not nil.

gupsho commented:
> @gupsho Would you please merge `master` into your branch?

Done.
